### PR TITLE
next/440/20240515/v1

### PR DESCRIPTION
**Pull request by victorjulien**
Staging:
 - #11062
 - #11063

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1832

### CODE REVIEW

**Review by jasonish — APPROVED** [review]

File: doc/userguide/rules/app-layer.rst [hunk 1]
```diff
@@ -8,18 +8,37 @@ Match on the detected app-layer protocol.
...
 A special value 'failed' can be used for matching on flows in which
 protocol detection failed. This can happen if Suricata doesn't know
 the protocol or when certain 'bail out' conditions happen.
 
+The different modes are
+* direction : protocol recognized on the direction of the current packet
+* to_server : protocol recognized in the direction to server
+* to_client : protocol recognized in the direction to client
+* either : tries to match protocols found on both directions
+* final : final protocol chosen by Suricata for parsing
```
Comment (jasonish) [inline doc/userguide/rules/app-layer.rst:31]: This just sinking in now. I think `final` needs more explanation.  For example, whats the difference between:
- `http,final`
- and just `http`

**Review by jasonish — APPROVED** [review]
For review, looks OK.

@catenacyber Can you look at adding a little more detail around final?

File: doc/userguide/rules/app-layer.rst (hunk 1 above)
Comment (catenacyber) [reply doc/userguide/rules/app-layer.rst:31]: see below

> By default, (if no mode is specified), the mode is ``direction``.

Comment (catenacyber) [reply doc/userguide/rules/app-layer.rst:31]: I guess the default is not your expectation, but that is the history, and the purpose of this PR is to provide more choices

File: doc/userguide/rules/app-layer.rst [hunk 2]
```diff
@@ -8,18 +8,37 @@ Match on the detected app-layer protocol.
...
+* original : original protocol (in case of protocol change)
+
+By default, (if no mode is specified), the mode is ``direction``.
+
+Here is an example of a rule matching non-http traffic on port 80:
+
+.. container:: example-rule
+
+    alert tcp any any -> any 80 (msg:"non-HTTP traffic over HTTP standard port"; flow:to_server; app-layer-protocol:!http,final; sid:1; )
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply doc/userguide/rules/app-layer.rst:40]: > or if is even really needed.

That depends on your intention...

This rule would trigger if you use a http client against a server on port 80 with another protocol (recognized by suricata)

**Comment by catenacyber** [comment]
> Can you look at adding a little more detail around final?

Do you have a better wording proposal ? I am not so good at these things...

**Comment by jasonish** [comment]
> Do you have a better wording proposal ? I am not so good at these things...

No, I don't understand it enough.  Guess I'm just after more discussion in the docs around its usage to help me understand more.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20641

**Pull request merged**